op-service/clock/loop.go

```diff
@@ -0,0 +1,62 @@
+package clock
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+// LoopFn is a simple ticker-loop with io.Closer support.
+// Note that ticks adapt; slow function calls may result in lost ticks.
+type LoopFn struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	ticker  Ticker
+	fn      func(ctx context.Context)
+	onClose func() error
+
+	wg sync.WaitGroup
+}
+
+// Close cancels the context of the ongoing function call, waits for the call to complete, and cancels further calls.
+// Close is safe to call again or concurrently. The onClose callback will be called for each Close call.
+func (lf *LoopFn) Close() error {
+	lf.cancel()  // stop any ongoing function call, and close the main loop
+	lf.wg.Wait() // wait for completion
+	if lf.onClose != nil {
+		return lf.onClose() // optional: user can specify function to close resources with
+	}
+	return nil
+}
+
+func (lf *LoopFn) work() {
+	defer lf.wg.Done()
+	defer lf.ticker.Stop() // clean up the timer
+	for {
+		select {
+		case <-lf.ctx.Done():
+			return
+		case <-lf.ticker.Ch():
+			ctx, cancel := context.WithCancel(lf.ctx)
+			lf.fn(ctx)
+			cancel()
```

Comment:
What's the purpose of deriving a local context here? Why not just
```suggestion
			lf.fn(lf.ctx)
```
?

And if there is a purpose, I'd make sure that a panic in `lf.fn` doesn't drop a call to `cancel` like
```suggestion
			ctx, cancel := context.WithCancel(lf.ctx)
			func() {
				defer cancel()
				lf.fn(ctx)
			}()
```

Reply:
The local context here ensures that things that are opened in the function call don't stick around for longer than the function call, to avoid a looping function from creating lots of unclosed resources